src/clarity-icons/clarity-icons-element.ts

```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
 
     if (host.hasAttribute("shape")) {
-        generateIcon(host, host.getAttribute("shape"));
+        const shapeAttrValue = host.getAttribute("shape").split(/\s/)[0];
+
+        if (shapeAttrValue === this.currentShapeAttrVal) {
+            return;
+        }
+
+        this.currentShapeAttrVal = shapeAttrValue;
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
+    }
+
+    if (host.hasAttribute("title")) {
+        const titleAttrValue = host.getAttribute("title");
+
+        if (this.currentTitleAttrVal !== titleAttrValue) {
+            this.currentTitleAttrVal = titleAttrValue;
+        }
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
     }
 
     if (host.hasAttribute("size")) {
-        setIconSize(host, host.getAttribute("size"));
+        const sizeAttrValue = host.getAttribute("title");
+
+        if (this.currentSizeAttrVal !== sizeAttrValue) {
+            this.currentSizeAttrVal = sizeAttrValue;
+            setIconSize(host, host.getAttribute("size"));
+        }
     }
+
+    this.injectTemplate(host);
 };
 
 ClarityIconElement.prototype.attributeChangedCallback = function(attributeName: string, oldValue: string,
                                                                  newValue: string) {
     const host = this as HTMLElement;
 
+    if (!this.initialized) {
+        this.clrIconUniqId = "_clr_icon_" + clrIconId;
+        clrIconId++;
+        this.currentShapeAttrVal = "";
+        this.currentShapeTemplate = "";
+        this.currentTitleAttrVal = "";
+        this.currentSizeAttrVal = "";
+        this.initialized = true;
+    }
+
     if (attributeName === "shape") {
-        generateIcon(host, newValue);
+        this.currentShapeAttrVal = newValue.split(/\s/)[0];
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
+    }
+    if (attributeName === "title") {
+        this.currentTitleAttrVal = newValue;
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
     }
     if (attributeName === "size") {
         setIconSize(host, newValue);
     }
+
+    this.injectTemplate(host);
+};
+
+ClarityIconElement.prototype.injectTemplate = function(hostElement: any) {
+    if (this.currentTitleAttrVal) {
+        hostElement.setAttribute("aria-labelledby", this.clrIconUniqId);
```

Comment:
When called from the `connectedCallback`, it sets `aria-labelledby` to undefined because the id is not ready yet. See my comment on the initialization happening in the wrong place.